ocaml/xapi/repository.ml

```diff
@@ -583,3 +702,291 @@ let get_repository_handler (req : Http.Request.t) s _ =
+      raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
+    | ui -> ui
+  in
+  match Xml.parse_file xml_file_path with
+  | Xml.Element  ("updates", _, children) ->
+    List.filter_map (fun n ->
+        match n with
+        | Xml.Element ("update", attr, update_nodes) ->
+          let ty =
+            match List.assoc_opt "type" attr with
+            | Some ty -> ty
+            | None -> ""
+          in
+          let ui = List.fold_left (fun acc node ->
+              match node with
+              | Xml.Element ("id", _, [Xml.PCData v]) ->
+                { acc with id = v }
+              | Xml.Element ("url", _, [Xml.PCData v]) ->
+                { acc with url = v }
+              | Xml.Element ("special_info", _, [Xml.PCData v]) ->
+                { acc with spec_info = v }
+              | Xml.Element ("summary", _, [Xml.PCData v]) ->
+                { acc with summary = v }
+              | Xml.Element ("description", _, [Xml.PCData v]) ->
+                { acc with description = v }
+              | Xml.Element ("recommended_guidances", _, gs) ->
+                { acc with rec_guidances = (parse_guidances gs) }
+              | Xml.Element ("absolute_guidances", _, gs) ->
+                { acc with abs_guidances = (parse_guidances gs) }
+              | Xml.Element ("guidance_applicabilities", _, apps) ->
+                { acc with guidance_applicabilities = (parse_applicabilities apps) }
+              | _ -> acc
+            ) { empty_updateinfo with update_type = ty } update_nodes
+            |> assert_valid_updateinfo
+          in
+          debug "updateinfo: %s" (string_of_updateinfo ui);
+          Some ui
+        | _ -> None
+    ) children
+    |> List.map (fun updateinfo -> (updateinfo.id, updateinfo))
+  | _ ->
+    error "Failed to parse updateinfo.xml: missing <updates>";
+    raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
+  | exception e ->
+    error "Failed to parse updateinfo.xml: %s" (ExnHelper.string_of_exn e);
+    raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
+
+let parse_updateinfo ~hash =
+  let repo_dir = Filename.concat !Xapi_globs.local_pool_repo_dir !Xapi_globs.pool_repo_name in
+  let repodata_dir = Filename.concat repo_dir "repodata" in
+  let repomd_xml_path = Filename.concat repodata_dir "repomd.xml" in
+  let md = parse_repomd repomd_xml_path in
+  if hash <> md.checksum then
+    (error "Unexpected mismatch between XAPI DB and YUM DB. Need to do pool.sync-updates again.";
+     raise Api_errors.(Server_error (createrepo_failed, [])));
+  let updateinfo_xml_gz_path = Filename.concat repo_dir md.location in
+  match Sys.file_exists updateinfo_xml_gz_path with
+  | false ->
+    error "File %s doesn't exist" updateinfo_xml_gz_path;
+    raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
+  | true ->
+    with_updateinfo_xml updateinfo_xml_gz_path parse_updateinfo_xml_file
+
+let compare_segment s1 s2 =
```

Comment:
is `compare_version_strings` a better name?

Reply:
I think a comment how these version strings look like would help to document the code. It's difficult to say whether the code is correct or not without having some examples. Version strings are notorious for having surprising ordering properties to the point that I believe this could be its own module.

Reply:
More examples are added. And fixed a few bugs with respect to these examples. Thanks.